RecoLocalMuon/GEMSegment/plugins/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,320 @@
+      // try not to "cluster" the hits but to "chain" them;
+      // it does the clustering but also does a better job
+      // for inclined tracks (not clustering them together;
+      // crossed tracks would be still clustered together) 
+      // 22.12.09: In fact it is not much more different 
+      // than the "clustering", we just introduce another
+      // variable in the game - Z. And it makes sense 
+      // to re-introduce Y (or actually wire group mumber)
+      // in a similar way as for the strip number - see
+      // the code below.
+      bool goodToMerge  = isGoodToMerge(seeds[NNN], seeds[MMM]);
+      if(goodToMerge){
+        // merge chains!
+        // merge by adding seed NNN to seed MMM and erasing seed NNN
+
+        // add seed NNN to MMM (lower to larger number)
+        seeds[MMM].insert(seeds[MMM].end(),seeds[NNN].begin(),seeds[NNN].end());
+
+        // mark seed NNN as used
+        usedCluster[NNN] = true;
+        // we have merged a seed (NNN) to the highter seed (MMM) - need to contimue to
+        // next seed (NNN+1)
+        break;
+      }
+
+    }
+  }
+
+  // hand over the final seeds to the output
+  // would be more elegant if we could do the above step with
+  // erasing the merged ones, rather than the
+
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    if(usedCluster[NNN]) continue; //skip seeds that have been marked as used up in merging
+    rechits_chains.push_back(seeds[NNN]);
+  }
+
+  //***************************************************************
+
+  return rechits_chains;
+}
+
+bool ME0SegAlgoMM::isGoodToMerge(EnsembleHitContainer & newChain, EnsembleHitContainer & oldChain) {
+   for(size_t iRH_new = 0;iRH_new<newChain.size();++iRH_new){
+    int layer_new = newChain[iRH_new]->me0Id().layer();     
+    float phi_new = theEnsemble.first->toGlobal(newChain[iRH_new]->localPosition()).phi();
+    float eta_new = theEnsemble.first->toGlobal(newChain[iRH_new]->localPosition()).eta();
+    for(size_t iRH_old = 0;iRH_old<oldChain.size();++iRH_old){      
+      int layer_old = oldChain[iRH_old]->me0Id().layer();
+      float phi_old = theEnsemble.first->toGlobal(oldChain[iRH_old]->localPosition()).phi();
+      float eta_old = theEnsemble.first->toGlobal(oldChain[iRH_old]->localPosition()).eta();
+      // to be chained, two hits need to be in neighbouring layers...
+      // or better allow few missing layers (upto 3 to avoid inefficiencies);
+      // however we'll not make an angle correction because it
+      // worsen the situation in some of the "regular" cases 
+      // (not making the correction means that the conditions for
+      // forming a cluster are different if we have missing layers -
+      // this could affect events at the boundaries ) 
+      //to be chained, two hits need also to be "close" in phi and eta
+      bool layerRequirementOK = abs(layer_new-layer_old)<5;
+      bool phiRequirementOK = fabs(phi_old-phi_new) < dPhiChainBoxMax;
```

Comment:
done, plus removed also magic number 5 (= number of layers -1), which might hurt when changing the ME0Geometry to have more or fewer layers.